subxt/src/error/mod.rs

```diff
@@ -0,0 +1,161 @@
+/// variant. Prefer to use the [`Error<E>`] and [`Error`] aliases over
+/// using this type directly.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum Error {
+    /// Io error.
+    #[error("Io error: {0}")]
+    Io(#[from] std::io::Error),
+    /// Codec error.
+    #[error("Scale codec error: {0}")]
+    Codec(#[from] codec::Error),
+    /// Rpc error.
+    #[error("Rpc error: {0}")]
+    Rpc(#[from] RpcError),
+    /// Serde serialization error
+    #[error("Serde json error: {0}")]
+    Serialization(#[from] serde_json::error::Error),
+    /// Invalid metadata error
+    #[error("Invalid Metadata: {0}")]
+    InvalidMetadata(#[from] InvalidMetadataError),
+    /// Invalid metadata error
+    #[error("Metadata: {0}")]
+    Metadata(#[from] MetadataError),
+    /// Runtime error.
+    #[error("Runtime error: {0:?}")]
+    Runtime(DispatchError),
+    /// Error decoding to a [`crate::dynamic::Value`].
+    #[error("Error decoding into dynamic value: {0}")]
+    Decode(#[from] DecodeError),
+    /// Error encoding from a [`crate::dynamic::Value`].
+    #[error("Error encoding from dynamic value: {0}")]
+    Encode(#[from] EncodeError),
+    /// Transaction progress error.
+    #[error("Transaction error: {0}")]
+    Transaction(#[from] TransactionError),
+    /// Block related error.
+    #[error("Block error: {0}")]
+    Block(#[from] BlockError),
+    /// An error encoding a storage address.
+    #[error("Error encoding storage address: {0}")]
+    StorageAddress(#[from] StorageAddressError),
+    /// The bytes representing an error that we were unable to decode.
+    #[error("An error occurred but it could not be decoded: {0:?}")]
+    Unknown(Vec<u8>),
+    /// Other error.
+    #[error("Other error: {0}")]
+    Other(String),
+}
+
+impl From<&str> for Error {
+    fn from(error: &str) -> Self {
+        Error::Other(error.into())
+    }
+}
+
+impl From<String> for Error {
+    fn from(error: String) -> Self {
+        Error::Other(error)
+    }
+}
+
+impl From<DispatchError> for Error {
+    fn from(error: DispatchError) -> Self {
+        Error::Runtime(error)
+    }
+}
+
+/// An RPC error. Since we are generic over the RPC client that is used,
+/// the error is boxed and could be casted.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum RpcError {
+    // Dev note: We need the error to be safely sent between threads
+    // for `subscribe_to_block_headers_filling_in_gaps` and friends.
+    /// Error related to the RPC client.
+    #[error("RPC error: {0}")]
+    ClientError(Box<dyn std::error::Error + Send + Sync + 'static>),
+    /// The RPC subscription dropped.
+    #[error("RPC error: subscription dropped.")]
+    SubscriptionDropped,
+}
+
+/// Block error
+#[derive(Clone, Debug, Eq, thiserror::Error, PartialEq)]
+#[non_exhaustive]
+pub enum BlockError {
+    /// The block
+    #[error("Could not find a block with hash {0} (perhaps it was on a non-finalized fork?)")]
+    BlockHashNotFound(String),
```

Comment:
I think it's sufficient to have BlockError, no need to duplicate the enum variant as well

```suggestion
    HashNotFound(String),
```

Reply:
Good point; I prefer that; thanks!